File: iris-ld/iris-ld-proofs/src/main/java/city/smartb/iris/ldproof/serializer/VerifiableJsonLdDeserializer.java
package city.smartb.iris.ldproof.serializer;

import city.smartb.iris.ldproof.VerifiableJsonLd;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.deser.std.StdDeserializer;

import java.io.IOException;
import java.util.Map;

public class VerifiableJsonLdDeserializer extends StdDeserializer<VerifiableJsonLd> {

    public VerifiableJsonLdDeserializer() { this(null); }

    public VerifiableJsonLdDeserializer(Class<?> vc) {
        super(vc);
    }

    @Override
    public VerifiableJsonLd deserialize(JsonParser jp, DeserializationContext ctxt)
            throws IOException {
        JsonNode node = jp.getCodec().readTree(jp);

        ObjectMapper mapper = new ObjectMapper();

        Map<String, Object> result = mapper.convertValue(node, new TypeReference<Map<String, Object>>(){});

        return new VerifiableJsonLd(result);
    }
}

File: iris-ld/iris-ld-proofs/src/main/kotlin/city/smartb/iris/ld/ldproof/serializer/VerifiableJsonLdDeserializer.kt
package city.smartb.iris.ld.ldproof.serializer

import city.smartb.iris.ld.ldproof.VerifiableJsonLd
import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.deser.std.StdDeserializer
import java.io.IOException

class VerifiableJsonLdDeserializer @JvmOverloads constructor(vc: Class<*>? = null) :
    StdDeserializer<VerifiableJsonLd>(vc) {
    @Throws(IOException::class)
    override fun deserialize(jp: JsonParser, ctxt: DeserializationContext): VerifiableJsonLd {
        val node = jp.codec.readTree<JsonNode>(jp)
        val mapper = ObjectMapper()
        val result = mapper.convertValue(node, object : TypeReference<Map<String, Any>>() {})
        return VerifiableJsonLd(result)
    }
}
